src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,117 @@
+            size++;
+        } else {
+            resize();
+            arr[size] = value;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        if (index == arr.length || size >= arr.length) {
+            resize();
+        }
 
+        T[] newArr = (T[]) new Object[arr.length];
+        System.arraycopy(arr, 0, newArr, 0, size);
+        newArr[index] = value;
+        System.arraycopy(arr, index, newArr, index + 1, size - index);
+        System.arraycopy(newArr, 0, arr, 0, newArr.length);
+        if (index >= 0 && index <= size) {
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            this.add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        return arr[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        arr[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T[] newArr;
+        T result = null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        result = arr[index];
+        newArr = (T[]) new Object[arr.length];
+        System.arraycopy(arr, 0, newArr, 0, index);
+        System.arraycopy(arr, index + 1, newArr, index, arr.length - index - 1);
+        System.arraycopy(newArr, 0, arr, 0, arr.length);
+        size--;
+        return result;
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public T remove(T element) throws NoSuchElementException {
+        boolean found = false;
+
+        for (int i = 0; i < size; i++) {
+            if (arr[i] == element || (arr[i] != null && arr[i].equals(element))) {
+                remove(i);
+                return element;
+            }
+        }
+        throw new NoSuchElementException("Element " + element + " was not found");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        if (size == 0) {
+            return true;
+        }
         return false;
     }
+
+    public void resize() {
+        int size = this.size + (this.size / 2); // or indexOfLastCharacter >> 1
+        T[] newArr = (T[]) new Object[size];
+        System.arraycopy(arr, 0, newArr, 0, this.size);
+        arr = (T[]) new Object[size];
+        System.arraycopy(newArr, 0, arr, 0, size);
+    }
```

Comment:
The 'resize' method should be private as it is an internal utility method. Also, the resizing logic is incorrect. You create a new array with the new size but then immediately overwrite it with another new array of the same size, which is unnecessary. Use only one new array and copy the elements over.